fs/smb/client/truenas_streams.c

```diff
@@ -0,0 +1,633 @@
+		int err;
+		__le64 eof = cpu_to_le64(total_written);
+		err = SMB2_set_eof(xid, tcon, fid.persistent_fid,
+		    fid.volatile_fid, current->tgid, &eof);
+	}
+
+	SMB2_close(xid, tcon, fid.persistent_fid, fid.volatile_fid);
+	if (rc) {
+		return rc;
+	}
+
+	return 0;
+}
+
+static int
+read_stream(struct dentry *dentry, struct cifs_tcon *tcon,
+	    char *dst, size_t dstsz, unsigned int xid,
+	    const char *path, const char *stream, u32 *pused)
+{
+	int rc, buf_type = CIFS_NO_BUFFER;
+	struct cifs_fid fid;
+	struct cifs_io_parms io_parms;
+	struct smb2_file_all_info info;
+	unsigned int bytes_read = 0;
+	unsigned int rsize, total_read;
+	u32 to_read;
+
+	rc = do_stream_open(tcon, CIFS_SB(dentry->d_sb), STREAM_OPEN_READ, xid,
+	    path, stream, 0, &fid, &info);
+        if (rc) {
+		return rc;
+        }
+
+	if (dstsz == 0) {
+		SMB2_close(xid, tcon, fid.persistent_fid, fid.volatile_fid);
+		*pused = le64_to_cpu(info.EndOfFile) + streams_samba_compat_enabled;
+		return 0;
+	}
+
+	to_read = le64_to_cpu(info.EndOfFile);
+	if ((to_read >= XATTR_LARGE_SIZE_MAX) ||
+	    ((to_read + streams_samba_compat_enabled) > dstsz)) {
+		/*
+		 * The SMB protocol and MS-FSCC does not provide an upper-bound
+		 * on the maximum size of an alternate data stream. ReFS on
+		 * windows limits ADS max size to 64 KiB, MacOS has no
+		 * real limit on size of resource forks. ZFS doesn't
+		 * have an upper bound on xattr size (these get written as
+		 * files when they're too large), but since our APIs here
+		 * are somewhat terrible (no pwrite / pread equivalent) we
+		 * limit to XATTR_LARGE_SIZE_MAX to reduce maximum allocation
+		 * size that we need to perform.
+		 *
+		 * Typically an alternate data stream will be less than 1 KiB.
+		 */
+
+		SMB2_close(xid, tcon, fid.persistent_fid, fid.volatile_fid);
+		return -ERANGE;
+	}
+
+	rsize = tcon->ses->server->ops->negotiate_rsize(tcon,
+	    CIFS_SB(dentry->d_sb)->ctx);
+
+	for (total_read = 0; to_read > total_read; total_read += bytes_read) {
+		unsigned int len = min(to_read, rsize);
+		unsigned int retries = 0;
+		rc = -EAGAIN;
+		char *pbuf = dst + total_read;
+
+		io_parms = (struct cifs_io_parms) {
+			.persistent_fid = fid.persistent_fid,
+			.volatile_fid = fid.volatile_fid,
+			.pid = current->tgid,
+			.tcon = tcon,
+			.offset = total_read,
+			.length = len,
+		};
+
+		while ((rc == -EAGAIN) && ((retries += 1) < STREAM_MAX_RETRIES)) {
+			rc = SMB2_read(xid, &io_parms, &bytes_read, &pbuf,
+			    &buf_type);
+		}
+
+		// If we've read to end of stream, rc will be 0 and bytes_read 0
+		if (rc || (bytes_read == 0))
+			break;
+	}
+
+	if (streams_samba_compat_enabled) {
+		// Append terminating NULL to match Samba behavior
+		*(dst + total_read + 1) = '\0';
+	}
+
+	SMB2_close(xid, tcon, fid.persistent_fid, fid.volatile_fid);
+	if (rc == 0) {
+		if (streams_samba_compat_enabled) {
+			*pused = total_read + streams_samba_compat_enabled;
+			*(dst + total_read) = '\0';
```

Comment:
Don't you set the NULL twice and for some reason even at different offsets?  Should the first `if` be removed?